slicerator.py

```diff
@@ -88,80 +226,85 @@ def __iter__(self):
-                    nums = range(len(self))
-                    # This next line fakes up numpy's bool masking without
-                    # importing numpy.
-                    rel_indices = [x for x, y in zip(nums, key) if y]
-                    indices = _index_generator(rel_indices, abs_indices)
-                    new_length = sum(key)
-                    return Slicerator(self._ancestor, indices, new_length)
-            try:
-                new_length = len(key)
-            except TypeError:
-                # The key is a generator; return a plain old generator.
-                # Without knowing the length of the *key*,
-                # we can't give a Slicerator
-                # Also it cannot be checked if values are in range.
-                gen = (self[_k if _k >= 0 else _len + _k] for _k in key)
-                return gen
-            else:
-                # The key is a list of in-range values. Build another
-                # Slicerator, again deferring computation.
-                if any(_k < -_len or _k >= _len for _k in key):
-                    raise IndexError("Keys out of range")
-                rel_indices = ((_k if _k >= 0 else _len + _k) for _k in key)
-                indices = _index_generator(rel_indices, abs_indices)
-                return Slicerator(self._ancestor, indices, new_length)
+        if not (isinstance(key, slice) or
+                isinstance(key, collections.Iterable)):
+            return self._get(self._map_index(key))
         else:
-            if key < -_len or key >= _len:
-                raise IndexError("Key out of range")
-            try:
-                abs_key = self._indices[key]
-            except TypeError:
-                key = key if key >= 0 else _len + key
-                for _, i in zip(range(key + 1), self.indices):
-                    abs_key = i
-            return self._get(abs_key)
+            rel_indices, new_length = key_to_indices(key, len(self))
+            if new_length is None:
+                return (self[k] for k in rel_indices)
+            indices = _index_generator(rel_indices, self.indices)
+            return Slicerator(self._ancestor, indices,
+                              new_length, self.propagated_attrs)
+
+    def __getattr__(self, name):
+        # to avoid infinite recursion, always check if public field is there
+        if 'propagated_attrs' not in self.__dict__:
+            self.propagated_attrs = []
+        if name in self.propagated_attrs:
+            attr = getattr(self._ancestor, name)
+            if hasattr(attr, '_indexed'):
+                return SliceableAttribute(self, getattr(self._ancestor, name))
+            else:
+                return getattr(self._ancestor, name)
+        raise AttributeError
 
     def __getstate__(self):
         # When serializing, return a list of the sliced and processed data
+        # Any exposed attrs are lost.
         return [self._get(key) for key in self.indices]
```

Comment:
I'm not sure whether this is the right default behavior. A major use case for Slicerators is for sequences that are too large to fit in memory or very expensive to compute. So serialization should probably get its own issue.